src/Functions/URL/cutURLParameter.cpp

```diff
@@ -1,82 +1,178 @@
+                arguments[0].column->getName(), getName());
+    }
+
+    static void cutURL(ColumnString::Chars & data, String pattern, size_t prev_offset, size_t & cur_offset, size_t & offset_adjust)
+    {
         pattern += '=';
         const char * param_str = pattern.c_str();
         size_t param_len = pattern.size();
 
-        size_t prev_offset = 0;
-        size_t res_offset = 0;
+        const char * url_begin = reinterpret_cast<const char *>(&data[prev_offset]);
+        const char * url_end = reinterpret_cast<const char *>(&data[cur_offset - 2]);
+        const char * begin_pos = url_begin;
+        const char * end_pos = begin_pos;
 
-        for (size_t i = 0; i < offsets.size(); ++i)
+        do
         {
-            size_t cur_offset = offsets[i];
+            const char * query_string_begin = find_first_symbols<'?', '#'>(url_begin, url_end);
+            if (query_string_begin + 1 >= url_end)
+                break;
 
-            const char * url_begin = reinterpret_cast<const char *>(&data[prev_offset]);
-            const char * url_end = reinterpret_cast<const char *>(&data[cur_offset]) - 1;
-            const char * begin_pos = url_begin;
-            const char * end_pos = begin_pos;
+            const char * pos = static_cast<const char *>(memmem(query_string_begin + 1, url_end - query_string_begin - 1, param_str, param_len));
+            if (pos == nullptr)
+                break;
 
-            do
+            if (pos[-1] != '?' && pos[-1] != '#' && pos[-1] != '&')
             {
-                const char * query_string_begin = find_first_symbols<'?', '#'>(url_begin, url_end);
-                if (query_string_begin + 1 >= url_end)
-                    break;
-
-                const char * pos = static_cast<const char *>(memmem(query_string_begin + 1, url_end - query_string_begin - 1, param_str, param_len));
-                if (pos == nullptr)
-                    break;
-
-                if (pos[-1] != '?' && pos[-1] != '#' && pos[-1] != '&')
-                {
-                    pos = nullptr;
-                    break;
-                }
-
-                begin_pos = pos;
-                end_pos = begin_pos + param_len;
+                pos = nullptr;
+                break;
+            }
+
+            begin_pos = pos;
+            end_pos = begin_pos + param_len;
+
+            /// Skip the value.
+            while (*end_pos && *end_pos != '&' && *end_pos != '#')
+                ++end_pos;
+
+            /// Capture '&' before or after the parameter.
+            if (*end_pos == '&')
+                ++end_pos;
+            else if (begin_pos[-1] == '&')
+                --begin_pos;
+        } while (false);
+
+        size_t cut_length = end_pos - begin_pos;
+        cur_offset -= cut_length;
+        offset_adjust += cut_length;
+        data.erase(data.begin() + prev_offset + (begin_pos - url_begin), data.begin() + prev_offset+  (end_pos - url_begin));
```

Comment:
Hm, I thought once more about this and my previous comment to save a copy.

Right now, the result data array is initialized as a copy of the input data array (l. 146). Function `cutUrl()` is then called for every URL (l. 155/163). The `erase()` in l. 136 removes a URL parameter + value by copying all characters behind the end of the removed section to the beginning of the removed section (see `erase()` in Common/PODArray.h). Now imagine a large column with one million URLs. For the first URL, we'd need to copy one million - 1 URLs, for the second URL, we copy one million - 2 URLs etc. That is quite inefficient.

Maybe it is better to only reserve sufficient space in the output data +offset arrays and then repeatedly append to it on a URL-by-URL basis, i.e. 1. fetch an input URL, 2. remove parameters, 3. append the result - in a loop.

Sorry for noticing this quite late.

Reply:
Hm. I forgot about `erase()`" danger". I will implement 3rd variant that will 
1. append URL to the end of `res_data`, 
2. do replacement of this appended fragment of `res_data` (and still use `erase()` it will be OK because there would not be too much copying).
3.  then append another URL to `res_data` on the next loop iteration.

Reply:
@rschu1ze Finally I implemented 3rd variant as explained above. Please take a look.